### ISSUE TITLE
First stage alignment training failed when TMA_CEloss=True

### ISSUE THREAD

**Issue by auspicious3000** [yl4579/StyleTTS#56]
Hello,

I used the config under Models/LibriTTS/ with your code, except that I set TMA_CEloss=True. The model is trained on LibriTTS-460 for 80 epochs with a batch size of 32. However, the alignment training failed. Do you have any tips on how to make the training work in this case?

Thanks!
<img src="https://github.com/yl4579/StyleTTS/assets/9756284/d0c15755-4daa-460e-bf9c-241e3a875863" width=50% height=50%>
<img src="https://github.com/yl4579/StyleTTS/assets/9756284/f0532027-b7c3-4ffb-bccf-b3d5d7f37c59" width=50% height=50%>

**Comment by auspicious3000** [comment]
In addition, there is another problem when setting TMA_CEloss=**False**. 
In this case, `s2s_attn = F.softmax(s2s_attn_feat, dim=-1)`, the last dim is normalized, but for `s2s_attn_mono` the 2nd dim is normalized. 
When calculating the l1 loss `F.l1_loss(s2s_attn, s2s_attn_mono)`, it forces the soft attention alignment close to its non-differentiable monotonic version. However, the soft attention has the last dim normalized while the monotonic attention has the 2nd dim normalized. If I understood correctly, this makes the comparison somewhat invalid.

**Comment by yl4579** [comment]
This is why I set `TMA_CEloss=False` by default even though theoretically `TMA_CEloss=False` is the correct way of doing it. The results I presented in the paper were also based on `TMA_CEloss=False`, because I found that `TMA_CEloss=True` is very unstable (I was only able to train it successfully on LJSpeech but not other datasets). One way of fixing it is by adjusting the weight of `loss_mono` but the reconstructed speech isn't better than `TMA_CEloss=False` in my experience. 

As for the loss concern, you are absolutely correct that the alignment when setting `TMA_CEloss=False` is technically not the right way of doing it (i.e., it aligns texts to mels rather than mels to texts), but this way is much more stable with `s2s_loss` than `TMA_CEloss=True` because they are optimizing the same objective and they produce better results than otherwise, even better than alignments used in VITS which is trained from scratch.
